File: kotlin/ktsearch/src/test/kotlin/ktsearch/SearchSettingsTest.kt
package ktsearch

import ktfind.addExtensions
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * @author cary on 7/30/16.
 */
class SearchSettingsTest {
    @Test
    fun testDefaultSettings() {
        val settings = getDefaultSettings()
        assertFalse(settings.archivesOnly)
        assertTrue(settings.colorize)
        assertFalse(settings.debug)
        assertFalse(settings.firstMatch)
        assertFalse(settings.includeHidden)
        assertEquals(0, settings.linesAfter)
        assertEquals(0, settings.linesBefore)
        assertEquals(150, settings.maxLineLength)
        assertFalse(settings.multiLineSearch)
        assertFalse(settings.printDirs)
        assertFalse(settings.printFiles)
        assertFalse(settings.printLines)
        assertFalse(settings.printResults)
        assertFalse(settings.printUsage)
        assertFalse(settings.printVersion)
        assertFalse(settings.searchArchives)
        assertTrue(settings.recursive)
        assertFalse(settings.uniqueLines)
        assertFalse(settings.verbose)
    }

    @Test
    fun testAddExtensions() {
        val settings = getDefaultSettings().copy(inExtensions = setOf("java", "scala"))
        assertEquals(2, settings.inExtensions.size.toLong())
        assertTrue(settings.inExtensions.contains("java"))
        assertTrue(settings.inExtensions.contains("scala"))
    }

    @Test
    fun testAddExtensionsAsCommaString() {
        val defaultSettings = getDefaultSettings()
        val settings = defaultSettings.
                copy(inExtensions = addExtensions("java,scala", defaultSettings.inExtensions))
        assertEquals(2, settings.inExtensions.size.toLong())
        assertTrue(settings.inExtensions.contains("java"))
        assertTrue(settings.inExtensions.contains("scala"))
    }

    @Test
    fun testSetArchivesOnly() {
        val settings = setArchivesOnly(getDefaultSettings(), true)
        assertTrue(settings.archivesOnly)
        assertTrue(settings.searchArchives)
    }
}

File: scala/scalasearch/src/test/scala/scalasearch/SearchSettingsTest.scala
package scalasearch

import org.scalatest.BeforeAndAfterAll
import org.scalatest.funsuite.AnyFunSuite

class SearchSettingsTest extends AnyFunSuite with BeforeAndAfterAll {

  def assertDefaultSettings(settings:SearchSettings): Unit = {
    assert(settings.archivesOnly == DefaultSearchSettings.archivesOnly)
    assert(settings.debug == DefaultSearchSettings.debug)
    assert(settings.firstMatch == DefaultSearchSettings.firstMatch)
    assert(settings.followSymlinks == DefaultSearchSettings.followSymlinks)
    assert(settings.includeHidden == DefaultSearchSettings.includeHidden)
    assert(settings.linesAfter == DefaultSearchSettings.linesAfter)
    assert(settings.linesBefore == DefaultSearchSettings.linesBefore)
    assert(settings.multiLineSearch == DefaultSearchSettings.multiLineSearch)
    assert(settings.printDirs == DefaultSearchSettings.printDirs)
    assert(settings.printFiles == DefaultSearchSettings.printFiles)
    assert(settings.printLines == DefaultSearchSettings.printLines)
    assert(settings.printResults == DefaultSearchSettings.printResults)
    assert(settings.printUsage == DefaultSearchSettings.printUsage)
    assert(settings.printVersion == DefaultSearchSettings.printVersion)
    assert(settings.searchArchives == DefaultSearchSettings.searchArchives)
    assert(settings.uniqueLines == DefaultSearchSettings.uniqueLines)
    assert(settings.verbose == DefaultSearchSettings.verbose)
  }

  // test defaults
  test("""test default settings""") {
    val settings = SearchSettings()
    assertDefaultSettings(settings)
  }

  // test SettingsBuilder
  test("""test SettingsBuilder""") {
    val settings = SearchSettings(archivesOnly = true, debug = true,
      inExtensions = Set("java", "scala"), searchPatterns = Set("Search".r))
    assert(settings.archivesOnly)
    //assert(settings.searchArchives)
    assert(settings.debug)
    //assert(settings.verbose)
    assert(settings.inExtensions.size == 2)
    assert(settings.inExtensions.contains("java"))
    assert(settings.inExtensions.contains("scala"))
    assert(settings.searchPatterns.size == 1)
    assert(settings.searchPatterns.toList.head.toString == "Search")
  }
}
